### ISSUE TITLE
Legion T5 28IMB05 support

### ISSUE THREAD

**Issue by CE1CECL** [4JX/L5P-Keyboard-RGB#61]
Ref: https://github.com/BartoszCichecki/LenovoLegionToolkit/issues/69 & https://github.com/ViRb3/LenovoController/issues/4

I don’t know if this project is for desktop Legions either, but there isn’t any program available but vantage, especially for Linux, you can’t run vantage on linux (vantage isn’t even open sourced by Lenovo, at least the SDK for lighting or whatever isn’t)

**Comment by 4JX** [comment]
This program only covers the RGB aspect of the Vantage app.

I don't think it would be very feasible to add support for a desktop with the app in it's current state, as most of the code is hard coded to expect the layout of a 4 zone keyboard. That is, unless they also provided you with one of those?

**Comment by CE1CECL** [comment]
> This program only covers the RGB aspect of the Vantage app.
> 
> I don't think it would be very feasible to add support for a desktop with the app in it's current state, as most of the code is hard coded to expect the layout of a 4 zone keyboard. That is, unless they also provided you with one of those?

![Capture](https://user-images.githubusercontent.com/52040854/167253435-f672f9c2-11a7-4c07-8bb3-0e9b2b343385.PNG)

This is what it looks like.

**Comment by CE1CECL** [comment]
Its also worth mentioning that opening up  v0.14.0 doesn't show anything, anything newer shows an error.

![Capture](https://user-images.githubusercontent.com/52040854/167253675-982fcd9b-9a08-4940-9b56-ffa3c9805adf.PNG)

**Comment by 4JX** [comment]
> Its also worth mentioning that opening up v0.13.1 doesn't show anything, anything newer shows an error.

Older versions silently failed/showed the error in the console.

Can you send screenshots of the configuration of the rest of components (the different sub menus in that 3/4 tab)? And do any of them have addressable colours or are they all preset effects?

**Comment by CE1CECL** [comment]
![1Capture](https://user-images.githubusercontent.com/52040854/167259429-e9aa1419-d9c0-4e2a-964c-a2ab8f27de96.PNG)
![2Capture](https://user-images.githubusercontent.com/52040854/167259430-7c6de459-4869-4999-aee4-a601a1dfd8a4.PNG)
![3Capture](https://user-images.githubusercontent.com/52040854/167259431-8488bfd7-f19a-4fb1-aae8-6e760e2d9e8c.PNG)
![4Capture](https://user-images.githubusercontent.com/52040854/167259433-c1b301ba-86ea-4ced-8926-b1cca58d85cd.PNG)
![5Capture](https://user-images.githubusercontent.com/52040854/167259434-3c2aa288-1303-4cc6-bce7-988e9600e6db.PNG)
![6Capture](https://user-images.githubusercontent.com/52040854/167259435-6e5e5967-0559-4949-bcbb-efa3023a81cb.PNG)
![7Capture](https://user-images.githubusercontent.com/52040854/167259436-1cde8c63-dd8f-4e60-9d76-ff2215f8d85d.PNG)
![8Capture](https://user-images.githubusercontent.com/52040854/167259437-f80fe47d-a460-4c21-94ad-37f074081b7f.PNG)
![9Capture](https://user-images.githubusercontent.com/52040854/167259439-4cfff16d-dd9e-4350-9a63-1f224009357f.PNG)
![Capture](https://user-images.githubusercontent.com/52040854/167259440-c73e6033-a04b-4476-8347-4110791f216c.PNG)

**Comment by CE1CECL** [comment]
If you aren't able to fix this (which i don't know if it will, desktop Legion's don't seem too common), do you think if i run Linux, I can pass through the RGB driver using KVM? Have you or anyone else tried?

**Comment by 4JX** [comment]
> If you aren't able to fix this (which i don't know if it will, desktop Legion's don't seem too common), do you think if i run Linux, I can pass through the RGB driver using KVM? Have you or anyone else tried?

Perhaps, haven't tried myself though.

The main issue with something like this is time. Your run-of-the-mill company will do *anything but* give developers any sort of help in developing their own solutions, because god forbid end users can enjoy their hardware after companies move on to their next big thing.

I was able to track down where the magic is probably happening, at least for my system:
![i](https://user-images.githubusercontent.com/79868816/167273991-8c27690d-a789-43bd-940c-a17970f00e64.png)

- `C:\ProgramData\Lenovo\ImController\Plugins\LenovoGamingSystemPlugin\x64\LedSetModule.dll` is the C# bridge for the led driver that interacts with vantage
- In that same folder, `LED.dll`, `LedDtSettingsModule.dll` and `LedSettingsModule.dll` seem to be the C/C++ part that actually does the job.

Specifically, `LedDtSettingsModule.dll` has strings that:
- Hint to GPUs (`..\GPU\AMDRt6.json`, `..\GPU\NvdiaLedRTX30.json`)
- Reference `Legion T530-28ICB`, `Legion T530-28APR` (both desktops) and, relevant to your case, the `DT550`
  - Going by the convention that the 5 Pro I'm using falls under the `NBY550` name which we can assume is "**N**ote**B**ook Y550", *and* with the folder where my current RGB settings are saved being at `C:\ProgramData\Lenovo\ImController\SystemPluginData\LenovoGamingSystemPlugin\data\ledprofile\NBY550`
  - `DT550` is most likely "**D**esktop T550", which [looks an awful lot like the japanese model name for your machine](https://taxandirect.jp/SHOP/90RCCTO1WW.html)

Its also worth noting that `LedSetModule.dll` has references to what probably are the different desktop build configurations:
![a](https://user-images.githubusercontent.com/79868816/167275009-3b344e1f-cd99-4f0e-8280-45821708fc44.png)

To sum up the information dump: Successfully reverse engineering `LedDtSettingsModule.dll` would probably be enough to allow for custom controls on the entire lineup of Lenovo (Gaming/Legion) machines. As for when and how, we go back to the time (and feature creep) problem.

**Comment by CE1CECL** [comment]
[C.zip](https://github.com/4JX/L5P-Keyboard-RGB/files/8648114/C.zip)

**Comment by 4JX** [comment]
Putting this on hold for now since it's going to take quite a bit of effort to make progress for something like this, apologies.

**Comment by CE1CECL** [comment]
@4JX OpenRGB works with my fan now, at least on windows 10. But the LED strip idk how to configure that, its doing this over HID VID 17EF & PID C955

**Comment by 4JX** [comment]
Been thinking about this issue for a while, I decided it would be better to not attempt to implement this out of concerns for feature creep as the program is focused on 4 zone keyboards.
